packages/js/src/utils/typescript/run-type-check.ts

```diff
@@ -0,0 +1,163 @@
+import * as path from 'path';
+import * as ts from 'typescript';
+import { existsSync, readFileSync } from 'fs';
+import * as chalk from 'chalk';
+import { codeFrameColumns } from '../code-frames/code-frames';
+
+export interface TypeCheckResult {
+  warnings?: string[];
+  errors?: string[];
+  inputFilesCount: number;
+  totalFilesCount: number;
+  incremental: boolean;
+}
+
+type TypeCheckOptions = BaseTypeCheckOptions & Mode;
+
+interface BaseTypeCheckOptions {
+  ts: typeof import('typescript');
+  workspaceRoot: string;
+  projectRoot: string;
+  tsConfigPath: string;
+  cacheDir?: string;
+}
+
+type Mode = NoEmitMode | EmitDeclarationOnlyMode;
+
+interface NoEmitMode {
+  mode: 'noEmit';
+}
+
+interface EmitDeclarationOnlyMode {
+  mode: 'emitDeclarationOnly';
+  outDir: string;
+}
+
+export async function runTypeCheck(
+  options: TypeCheckOptions
+): Promise<TypeCheckResult> {
+  const { ts, workspaceRoot, projectRoot, tsConfigPath, cacheDir } = options;
+  const config = ts.readConfigFile(
+    path.join(workspaceRoot, tsConfigPath),
```

Comment:
suggestion: can we make sure the consumers of `runTypeCheck` pass in the correct `tsConfigPath` so that `runTypeCheck` does not have to do this `join(...)`. React and Node (and Node w/ external dependencies) behave different when it comes to `tsconfig` (w/ the possibility of `tsconfig.generated` being used instead)?

Reply:
Yeah we can change that :)